funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/result_queue_subscriber.py

```diff
@@ -0,0 +1,342 @@
+        self._channel.queue_bind(
+            queue=self.queue_name,
+            exchange=self.exchange,
+            routing_key=self.routing_key,
+            callback=self._on_bindok,
+        )
+
+    def _on_bindok(self, unused_frame):
+        """Invoked by pika when the Queue.Bind method has completed. At this
+        point we will start consuming messages by calling start_consuming
+        which will invoke the needed RPC commands to start the process.
+
+        :param pika.frame.Method unused_frame: The Queue.BindOk response frame
+
+        """
+        logger.info("Queue bound")
+        self.start_consuming()
+
+    def start_consuming(self):
+        """This method sets up the consumer by first calling
+        add_on_cancel_callback so that the object is notified if RabbitMQ
+        cancels the consumer. It then issues the Basic.Consume RPC command
+        which returns the consumer tag that is used to uniquely identify the
+        consumer with RabbitMQ. We keep the value to use it when we want to
+        cancel consuming. The on_message method is passed in as a callback pika
+        will invoke when a message is fully received.
+
+        """
+        self.add_on_cancel_callback()
+        logger.debug("Waiting for basic_consume")
+        self._channel.basic_consume(
+            self.queue_name, on_message_callback=self.on_message, exclusive=True
+        )
+
+    def add_on_cancel_callback(self):
+        """Add a callback that will be invoked if RabbitMQ cancels the consumer
+        for some reason. If RabbitMQ does cancel the consumer,
+        on_consumer_cancelled will be invoked by pika.
+
+        """
+        logger.info("Adding consumer cancellation callback")
+        self._channel.add_on_cancel_callback(self.on_consumer_cancelled)
+
+    def on_consumer_cancelled(self, method_frame):
+        """Invoked by pika when RabbitMQ sends a Basic.Cancel for a consumer
+        receiving messages.
+
+        :param pika.frame.Method method_frame: The Basic.Cancel frame
+
+        """
+        logger.info("Consumer was cancelled remotely, shutting down: %r", method_frame)
+        if self._channel:
+            self._channel.close()
+
+    def on_message(self, _unused_channel, basic_deliver, _properties, body):
+        """on_message pushes a message upon receipt into the external_queue
+        for async consumption. The message pushed to the external_queue
+        is of the following type : Tuple[str, bytes]
+
+        :param pika.channel.Channel unused_channel: The channel object
+        :param pika.Spec.Basic.Deliver: basic_deliver method
+        :param pika.Spec.BasicProperties: properties
+        :param str|unicode body: The message body
+
+        """
+
+        routing_key = basic_deliver.routing_key
+
+        if self.external_queue:
+            try:
+                response: Tuple[str, bytes] = (routing_key.split(".")[0], body)
+                self.external_queue.put(response)
```

Comment:
I'm open to changing this, this probably should be `routing_key.rsplit(".", 1)[0]` to just remove the `.results` suffix.

Reply:
I still think this should not be splitting on the routing key at all. This method will get triggered on any received message. It's better not to make an easily avoidable assumption here (one which will break if we have messages in the future with no `.` in the key).

Also, I just noticed the use of `Tuple[str, bytes]`. Now that we're on py3.7+ it's almost always better to `from __future__ import annotations` at the top of the file to get deferred annotation support. Then you can write `tuple[str, bytes]` (builtin `tuple`). Because the annotation evaluation is deferred, this works fine even though class getitem on builtin collection types wasn't implemented until py3.9 .